src/contracts/GPv2Settlement.sol

```diff
@@ -137,6 +137,119 @@ contract GPv2Settlement is GPv2Signing, ReentrancyGuard, StorageAccessible {
         emit Settlement(msg.sender);
     }
 
+    /// @dev Settle a single order directly with on-chain liquidity. This
+    /// function is provided as a "fast-path" for settlements without any
+    /// coincidence of wants.
+    ///
+    /// This type of settlement always assumes the order will be filled in full,
+    /// and while it does accept partially fillable orders, they must have no
+    /// prior filled amount, and will be executed in full.
+    ///
+    /// @param order The single order to settle.
+    /// @param signingScheme The signing scheme used for the order.
+    /// @param signature The signature of the order used to recover the owner
+    /// address.
+    /// @param transfers Direct transfers of user funds to execute to interact
+    /// with on-chain liquidity.
+    /// @param interactions Smart contract interaction to perform with executed
+    /// sell amount of the order.
+    function settleOrder(
+        GPv2Order.Data memory order,
+        Scheme signingScheme,
+        bytes calldata signature,
+        GPv2AllowanceManager.Transfer[] calldata transfers,
+        GPv2Interaction.Data[] calldata interactions
+    ) external nonReentrant onlySolver {
+        GPv2TradeExecution.Data memory executedTrade;
+        bytes memory orderUid;
+        {
```

Comment:
The difficulty there is that the normal settlement relies on `calldata` structures and data. This means we would have to pass in a `Trade` and a token list and would be slightly more inefficient (as it would have to extract the order from trade one field at a time instead of the compiler generating a `CALLDATACOPY` instruction to get the order in one go, not to mention the token addresses would be encoded in the "token list" with the trade containing sell and buy token indices to recover this information).

To be completely honest, I didn't even try this approach, maybe the gas costs aren't that bad after all.

Reply:
I'd also like to see this tested out. I'd like to see recoverOrderFromTrade used to avoid that we create some assumptions there that are not true here (for example, before an order could only be sell or buy, now any `bytes32` works) and to avoid code duplication (computing orderUid).

Reply:
I completely agree. I think the code was quite ugly as it currently is (there is a lot of duplicated logic for example, and details like `receiver == address(0) ? owner : receiver` that are easy to forget. I was a bit over-eager to optimize :stuck_out_tongue: